### DISCUSSION TITLE
What exactly is a serverless function? and can it help me track email openings when I send email?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#67964] [Help] [answered]
### Summary

I have Nodemailer to send emails, and I am using their DSN status notifications for email delivery, but it doesn't work to see if an email has been opened.

What exactly is a serverless function? and could I implemented to check if somebody has opened an email? or say unsubscribe from an email list?

Another way to track for what I have been reading, is to send an image with a src for specific route in your server, but what if I have 1,000+ emails?, Now I would need to create one route per email, how would this work in Next.js?

**Comment by icyJoseph** [comment, marked as answer]
A serverless function, in a nutshell, in 2024:

1. Controlled by a runtime
2. The runtime knows which code to execute when the function is called, so it loads that (this takes a given amount of time)
3. The runtime executes the function
4. There might be a bit of a grace time period where the function is kept around, to avoid re-doing 2
5. The runtime kills the function

To check if an email has been opened, 🤔 I haven't looked at this back in 2014... I guess, you can add an image or some kind of asset, with a unique URL, and see if that URL is hit. 

I would assume there are other techniques and such, but my knowledge is limited in that area.

> Another way to track for what I have been reading, is to send an image with a src for specific route in your server, but what if I have 1,000+ emails?, Now I would need to create one route per email, how would this work in Next.js?

Not really, just one endpoint can handle all of this. Either by looking at the queries that arrive, or generating a dynamic path: https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes

This doesn't tell you the email was read though, only that the resource was somehow loaded.

**Reply by Elindo586** [reply]
Apparently you can include parameters in the web address, from there req.params and do whatever else you want.

Gonna be testing something like:

```
http://127.0.0.1:5001/api/users/123


app.get('/api/users/:id', (req, res) => {
  const userId = req.params.id;
  // Do something with userId
});
```
